planner/core/rule_generate_column_substitute.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"context"
+
+	"github.com/pingcap/parser/ast"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/sessionctx/stmtctx"
+	"github.com/pingcap/tidb/types"
+)
+
+type gcSubstituter struct {
+}
+
+// ExprColumnMap is used to store all expressions of indexed generated columns in a table,
+// and map them to the generated columns,
+// thus we can substitute the expression in a query to an indexed generated column.
+type ExprColumnMap map[expression.Expression]*expression.Column
+
+func (gc *gcSubstituter) optimize(ctx context.Context, lp LogicalPlan) (LogicalPlan, error) {
+	exprToColumn := make(ExprColumnMap)
+	collectGenerateColumn(lp, exprToColumn)
+	if len(exprToColumn) == 0 {
+		return lp, nil
+	}
+	return gc.substitute(ctx, lp, exprToColumn, lp.SCtx().GetSessionVars().StmtCtx), nil
+}
+
+// collectGenerateColumn collect the generate column and save them to a map from their expressions to themselves.
+// For the sake of simplicity, we don't collect the stored generate column because we can't their expressions directly.
+// TODO: support stored generate column.
+func collectGenerateColumn(lp LogicalPlan, exprToColumn ExprColumnMap) {
+	if ds, ok := lp.(*DataSource); ok {
+		tblInfo := ds.tableInfo
+		for _, idx := range tblInfo.Indices {
+			for _, idxPart := range idx.Columns {
+				colInfo := tblInfo.Columns[idxPart.Offset]
+				if colInfo.IsGenerated() && !colInfo.GeneratedStored {
+					s := ds.schema.Columns
+					col := expression.ColInfo2Col(s, colInfo)
+					if col != nil {
+						exprToColumn[col.VirtualExpr] = col
+					}
+				}
+			}
+		}
+		return
+	}
+	for _, child := range lp.Children() {
+		collectGenerateColumn(child, exprToColumn)
+	}
+}
+
+func (gc *gcSubstituter) substitute(ctx context.Context, lp LogicalPlan, exprToColumn ExprColumnMap, sctx *stmtctx.StatementContext) LogicalPlan {
+	var expr *expression.Expression
```

Comment:
Why using `*expression.Expression`? Can we just use an `expression.Expression` here?

Reply:
Then we can't substitue it